File: packages/openmemory-js/src/sources/google_sheets.ts
/**
 * google sheets source for openmemory - production grade
 * requires: googleapis
 * env vars: GOOGLE_SERVICE_ACCOUNT_FILE or GOOGLE_CREDENTIALS_JSON
 */

import { base_source, source_config_error, source_item, source_content } from './base';
import type { sheets_v4, google as google_type } from 'googleapis';

export class google_sheets_source extends base_source {
    override name = 'google_sheets';
    private service: sheets_v4.Sheets | null = null;
    private auth: any = null;

    async _connect(creds: Record<string, any>): Promise<boolean> {
        let google_mod: typeof google_type;
        try {
            google_mod = await import('googleapis').then(m => m.google);
        } catch {
            throw new source_config_error('missing deps: npm install googleapis', this.name);
        }

        const scopes = ['https://www.googleapis.com/auth/spreadsheets.readonly'];

        if (creds.credentials_json) {
            this.auth = new google_mod.auth.GoogleAuth({ credentials: creds.credentials_json, scopes });
        } else if (creds.service_account_file) {
            this.auth = new google_mod.auth.GoogleAuth({ keyFile: creds.service_account_file as string, scopes });
        } else if (process.env.GOOGLE_CREDENTIALS_JSON) {
            this.auth = new google_mod.auth.GoogleAuth({ credentials: JSON.parse(process.env.GOOGLE_CREDENTIALS_JSON), scopes });
        } else if (process.env.GOOGLE_SERVICE_ACCOUNT_FILE) {
            this.auth = new google_mod.auth.GoogleAuth({ keyFile: process.env.GOOGLE_SERVICE_ACCOUNT_FILE, scopes });
        } else {
            throw new source_config_error('no credentials: set GOOGLE_SERVICE_ACCOUNT_FILE or GOOGLE_CREDENTIALS_JSON', this.name);
        }

        this.service = google_mod.sheets({ version: 'v4', auth: this.auth });
        return true;
    }

    async _list_items(filters: Record<string, any>): Promise<source_item[]> {
        if (!this.service) throw new source_config_error('not connected', this.name);
        if (!filters.spreadsheet_id) {
            throw new source_config_error('spreadsheet_id is required', this.name);
        }

        try {
            const meta = await this.service.spreadsheets.get({ spreadsheetId: filters.spreadsheet_id as string });

            return (meta.data.sheets || []).map((sheet: any, i: number) => ({
                id: `${filters.spreadsheet_id}!${sheet.properties?.title || 'Sheet1'}`,
                name: sheet.properties?.title || 'Sheet1',
                type: 'sheet',
                index: i,
                spreadsheet_id: filters.spreadsheet_id as string
            }));
        } catch (e: unknown) {
            const msg = e instanceof Error ? e.message : String(e);
            throw new Error(`[google_sheets] list failed: ${msg}`);
        }
    }

    async _fetch_item(item_id: string): Promise<source_content> {
        if (!this.service) throw new source_config_error('not connected', this.name);

        const [spreadsheet_id, sheet_range] = item_id.includes('!')
            ? item_id.split('!', 2)
            : [item_id, 'A:ZZ'];

        try {
            const result = await this.service.spreadsheets.values.get({
                spreadsheetId: spreadsheet_id,
                range: sheet_range
            });

            const values = result.data.values || [];

            // convert to markdown table
            const lines = values.map((row: any[], i: number) => {
                const line = row.map(String).join(' | ');
                return i === 0 ? `${line}\n${row.map(() => '---').join(' | ')}` : line;
            });

            const text = lines.join('\n');

            return {
                id: item_id,
                name: sheet_range,
                type: 'spreadsheet',
                text,
                data: text,
                meta: { source: 'google_sheets', spreadsheet_id, range: sheet_range, row_count: values.length }
            };
        } catch (e: unknown) {
            const msg = e instanceof Error ? e.message : String(e);
            throw new Error(`[google_sheets] fetch failed: ${msg}`);
        }
    }
}

File: packages/openmemory-py/src/openmemory/connectors/google_sheets.py
"""
google sheets connector for openmemory
requires: google-api-python-client, google-auth
env vars: GOOGLE_SERVICE_ACCOUNT_FILE or GOOGLE_CREDENTIALS_JSON
"""
from typing import List, Dict, Optional
import os
import json
from .base import BaseConnector, SourceItem, SourceContent, SourceAuthError

class GoogleSheetsConnector(BaseConnector):
    """connector for google sheets"""

    name = "google_sheets"

    def __init__(self, user_id: Optional[str] = None):
        super().__init__(user_id)
        self.service = None
        self.creds = None

    async def connect(self, **creds) -> bool:
        """
        authenticate with google sheets api
        
        env vars:
            GOOGLE_SERVICE_ACCOUNT_FILE: path to service account json
            GOOGLE_CREDENTIALS_JSON: raw json string of credentials
        """
        try:
            from google.oauth2 import service_account
            from googleapiclient.discovery import build
        except ImportError:
            raise ImportError("pip install google-api-python-client google-auth")

        scopes = ["https://www.googleapis.com/auth/spreadsheets.readonly"]

        if "credentials_json" in creds:
            self.creds = service_account.Credentials.from_service_account_info(
                creds["credentials_json"], scopes=scopes
            )
        elif "service_account_file" in creds:
            self.creds = service_account.Credentials.from_service_account_file(
                creds["service_account_file"], scopes=scopes
            )
        elif os.environ.get("GOOGLE_CREDENTIALS_JSON"):
            info = json.loads(os.environ["GOOGLE_CREDENTIALS_JSON"])
            self.creds = service_account.Credentials.from_service_account_info(info, scopes=scopes)
        elif os.environ.get("GOOGLE_SERVICE_ACCOUNT_FILE"):
            self.creds = service_account.Credentials.from_service_account_file(
                os.environ["GOOGLE_SERVICE_ACCOUNT_FILE"], scopes=scopes
            )
        else:
            raise ValueError("no google credentials provided")

        self.service = build("sheets", "v4", credentials=self.creds)
        self._connected = True
        return True

    async def list_items(
        self, spreadsheet_id: Optional[str] = None, **filters
    ) -> List[SourceItem]:  # type: ignore[override]
        """
        list sheets in a spreadsheet
        
        args:
            spreadsheet_id: the spreadsheet id to list sheets from
        """
        if not self._connected or not self.service:
            await self.connect()

        if not spreadsheet_id:
            raise ValueError("spreadsheet_id is required")

        if not self.service:
            raise SourceAuthError("Service not initialized", self.name)

        meta = self.service.spreadsheets().get(spreadsheetId=spreadsheet_id).execute()

        sheets: List[SourceItem] = []
        for sheet in meta.get("sheets", []):
            props = sheet.get("properties", {})
            sheets.append(
                SourceItem(
                    id=f"{spreadsheet_id}!{props.get('title', 'Sheet1')}",
                    name=props.get("title", "Sheet1"),
                    type="sheet",
                    path=f"/{spreadsheet_id}",
                    metadata={  # type: ignore[arg-type]  # type: ignore[arg-type]
                        "index": props.get("index", 0),
                        "spreadsheet_id": spreadsheet_id,
                    },
                )
            )

        return sheets

    async def fetch_item(self, item_id: str) -> SourceContent:  # type: ignore[override]
        """
        fetch sheet data as text
        
        item_id format: "spreadsheet_id!sheet_name" or just "spreadsheet_id"
        """
        if not self._connected:
            await self.connect()

        # parse item_id
        if "!" in item_id:
            spreadsheet_id, sheet_range = item_id.split("!", 1)
        else:
            spreadsheet_id = item_id
            sheet_range = "A:ZZ"  # all columns

        if not self.service:
            raise SourceAuthError("Service not initialized", self.name)

        result = self.service.spreadsheets().values().get(
            spreadsheetId=spreadsheet_id,
            range=sheet_range
        ).execute()

        values = result.get("values", [])

        # convert to text table
        lines = []
        for row in values:
            lines.append(" | ".join([str(cell) for cell in row]))

        text = "\n".join(lines)

        return SourceContent(
            id=item_id,
            name=sheet_range,
            type="spreadsheet",
            text=text,
            data=text,
            metadata={  # type: ignore[arg-type]  # type: ignore[arg-type]
                "source": "google_sheets",
                "spreadsheet_id": spreadsheet_id,
                "range": sheet_range,
                "row_count": len(values),
            },
        )
